src/protocols/fiat-protocol.js

```diff
@@ -0,0 +1,296 @@
+ */
+export class IFiatProtocol {
+  /**
+   * Gets a quote for a crypto asset purchase.
+   * @param {Omit<BuyOptions, 'recipient'>} options - The options for the buy operation.
+   * @returns {Promise<FiatQuote>} A quote for the transaction.
+   */
+  async quoteBuy (options) {
+    throw new NotImplementedError('quoteBuy(options)')
+  }
+
+  /**
+   * Generates a widget URL for a user to purchase a crypto asset with fiat currency.
+   * @param {BuyOptions} options - The options for the buy operation.
+   * @returns {Promise<BuyResult>} The operation's result.
+   */
+  async buy (options) {
+    throw new NotImplementedError('buy(options)')
+  }
+
+  /**
+   * Gets a quote for a crypto asset sale.
+   * @param {Omit<SellOptions, 'refundAddress'>} options - The options for the sell operation.
+   * @returns {Promise<FiatQuote>} A quote for the transaction.
+   */
+  async quoteSell (options) {
+    throw new NotImplementedError('quoteSell(options)')
+  }
+
+  /**
+   * Generates a widget URL for a user to sell a crypto asset for fiat currency.
+   * @param {SellOptions} options - The options for the sell operation.
+   * @returns {Promise<SellResult>} The operation's result.
+   */
+  async sell (options) {
+    throw new NotImplementedError('sell(options)')
+  }
+
+  /**
+   * Retrieves the details of a specific transaction from the provider.
+   * @param {string} txId - The unique identifier of the transaction.
+   * @returns {Promise<FiatTransactionDetail>} The transaction details.
+   */
+  async getTransactionDetail (txId) {
+    throw new NotImplementedError('getTransactionDetail(txId)')
+  }
+
+  /**
+   * Retrieves a list of supported crypto assets from the provider.
+   * @returns {Promise<SupportedCryptoAsset[]>} An array of supported crypto assets.
+   */
+  async getSupportedCryptoAssets () {
+    throw new NotImplementedError('getSupportedCryptoAssets()')
+  }
+
+  /**
+   * Retrieves a list of supported fiat currencies from the provider.
+   * @returns {Promise<SupportedFiatCurrency[]>} An array of supported fiat currencies.
+   */
+  async getSupportedFiatCurrencies () {
+    throw new NotImplementedError('getSupportedFiatCurrencies()')
+  }
+
+  /**
+   * Retrieves a list of supported countries from the provider.
+   * @returns {Promise<SupportedCountry[]>} An array of supported countries.
+   */
+  async getSupportedCountries () {
+    throw new NotImplementedError('getSupportedCountries()')
+  }
+}
+
+/**
+ * @abstract
+ * @implements {IFiatProtocol}
+ */
+export default class FiatProtocol {
+  /**
+   * Creates a new fiat protocol with read-only account.
+   *
+   * @overload
+   * @param {IWalletAccountReadOnly} [account] - The wallet account to use to interact with the protocol.
+   */
+
+  /**
+   * Creates a new fiat protocol with read-only account.
+   *
+   * @overload
+   * @param {IWalletAccount} [account] - The wallet account to use to interact with the protocol.
+   */
+  constructor (account = undefined) {
```

Comment:
This is slightly better:
```suggestion
  /**
   * Creates a new interface to the moon-pay protocol without binding it to a wallet account.
   *
   * @overload
   * @param {undefined} [account] - The wallet account to use to interact with the protocol.
   */

  /**
   * Creates a new read-only interface to the moon-pay protocol.
   *
   * @overload
   * @param {IWalletAccountReadOnly} account - The wallet account to use to interact with the protocol.
   */

  /**
   * Creates a new interface to the moon-pay protocol.
   *
   * @overload
   * @param {IWalletAccount} account - The wallet account to use to interact with the protocol.
   */
  constructor (account) {
```

Reply:
Since this is a generic interface that will support more providers that just Moonpay, can we omit that specific provider in the description, or make it clear Moonpay is an example of a provider that is compatible with IFiatProtocol?

Reply:
Ohhh, my bad sorry. That was not intended, i just copied and pasted the doc from the moon-pay protocol repo:
```suggestion
  /**
   * Creates a new interface to the protocol without binding it to a wallet account.
   *
   * @overload
   * @param {undefined} [account] - The wallet account to use to interact with the protocol.
   */
  /**
   * Creates a new read-only interface to the protocol.
   *
   * @overload
   * @param {IWalletAccountReadOnly} account - The wallet account to use to interact with the protocol.
   */
  /**
   * Creates a new interface to the protocol.
   *
   * @overload
   * @param {IWalletAccount} account - The wallet account to use to interact with the protocol.
   */
  constructor (account) {
```